File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
def solution(operations):
    answer = []
    heap = []
    heap_max = []
    for i in operations:
        op = i[0]
        if op == "I":#푸쉬
            num = int(i[2:])
            heapq.heappush(heap,num)
            heapq.heappush(heap_max,(-num,num))

        else:
            if not heap:
                continue
            else:
                op = i[2:]
                if op =='-1': #최솟값 삭제
                    temp = heapq.heappop(heap)
                    heap_max.remove((-temp,temp))
                else:
                    temp = heapq.heappop(heap_max)[1]
                    heap.remove(temp)
    if heap:
        answer=[heapq.heappop(heap_max)[1],heapq.heappop(heap)]
    else:
        answer = [0,0]
    return answer

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    static PriorityQueue<Integer> max = new PriorityQueue<>((o1, o2) -> o2 - o1);
    static PriorityQueue<Integer> min = new PriorityQueue<>();
    
    public int[] solution(String[] operations) {
        int[] answer = new int[2];
        for(String oper : operations){
            String[] temp = oper.split(" ");
            if(temp[0].equals("I")){
                //숫자 삽입
                int n = Integer.parseInt(temp[1]);
                max.offer(n);
                min.offer(n);
            }
            else{
                
                //모든 큐가 비어있다면 건너뛰기
                if(max.isEmpty()) continue;
                
                if(temp[1].equals("1")){
                    //최댓값 삭제
                    int n = max.poll();
                    min.remove(n);
                }
                else{
                    //최솟값 삭제
                    int n = min.poll();
                    max.remove(n);
                }
            }
        }
        if(!max.isEmpty()){
            answer[0] = max.peek();
            answer[1] = min.peek();
        }
        return answer;
    }
}